exchanges/coinbene/coinbene_websocket.go

```diff
@@ -0,0 +1,328 @@
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wshandler"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wsorderbook"
+)
+
+const (
+	coinbeneWsURL = "wss://ws-contract.coinbene.vip/openapi/ws"
+	event         = "event"
+	topic         = "topic"
+)
+
+// WsConnect connects to websocket
+func (c *Coinbene) WsConnect() error {
+	if !c.Websocket.IsEnabled() || !c.IsEnabled() {
+		return errors.New(wshandler.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	err := c.WebsocketConn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	go c.WsDataHandler()
+	if c.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		c.Login()
+	}
+	c.GenerateDefaultSubscriptions()
+
+	return nil
+}
+
+// GenerateDefaultSubscriptions generates stuff
+func (c *Coinbene) GenerateDefaultSubscriptions() {
+	var channels = []string{"orderBook.%s.100", "tradeList.%s", "ticker.%s", "kline.%s"}
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	for x := range channels {
+		for y := range c.EnabledPairs {
+			c.EnabledPairs[y].Delimiter = ""
+			subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+				Channel:  fmt.Sprintf(channels[x], c.EnabledPairs[y]),
+				Currency: c.EnabledPairs[y],
+			})
+		}
+	}
+	c.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// GenerateAuthSubs generates auth subs
+func (c *Coinbene) GenerateAuthSubs() {
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	var sub wshandler.WebsocketChannelSubscription
+	var userChannels = []string{"user.account", "user.position", "user.order"}
+	for z := range userChannels {
+		sub.Channel = userChannels[z]
+		subscriptions = append(subscriptions, sub)
+	}
+	c.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// WsDataHandler handles websocket data
+func (c *Coinbene) WsDataHandler() {
+	c.Websocket.Wg.Add(1)
+
+	defer c.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-c.Websocket.ShutdownC:
+			return
+
+		default:
+			stream, err := c.WebsocketConn.ReadMessage()
+			if err != nil {
+				c.Websocket.DataHandler <- err
+				return
+			}
+			c.Websocket.TrafficAlert <- struct{}{}
+			if string(stream.Raw) == "ping" {
+				c.WebsocketConn.Lock()
+				c.WebsocketConn.Connection.WriteMessage(websocket.TextMessage, []byte("pong"))
+				c.WebsocketConn.Unlock()
+				continue
+			}
+			var result map[string]interface{}
+			err = common.JSONDecode(stream.Raw, &result)
+			if err != nil {
+				c.Websocket.DataHandler <- err
+			}
+			_, ok := result[event]
+			switch {
+			case ok && (result[event].(string) == "subscribe" || result[event].(string) == "unsubscribe"):
+				continue
+			case ok && result[event].(string) == "error":
+				c.Websocket.DataHandler <- fmt.Errorf("message: %s. code: %v", result["message"], result["code"])
+				continue
+			}
+			if ok && strings.Contains(result[event].(string), "login") {
+				if result["success"].(bool) {
+					c.GenerateAuthSubs()
+				}
+				continue
```

Comment:
In the event of a login auth failure, set websocket authenticated support to disabled and report an auth error